Fixing HF inconsistencies

xdvriend opened the pull request:
**Short description**
The UHF `QCMethod` was still written using an `RSQHamiltonian`. his PR changes that. Now all HF methods use their own Hamiltonian.

Since there is another inconsistency regarding the overlap matrix, which should be an operator, I'll tackle that here as well.

**Related issues**
closes #840 
closes #845 

**Additional context**
I have also updated the python bindings and updated the example notebooks that use UHF calculations. They are now up to date with the latest python bindings. 

**Don't forget:**
- [x] if new files are created: update the Doxygen file, the collective gqcp.hpp header and the CMake files

lelemmen commented on gqcp/tests/QCMethod/HF/UHF/QCMethod_UHF_stability_test.cpp:
```diff
@@ -87,10 +85,10 @@ BOOST_AUTO_TEST_CASE(H4_stability_test) {
-    const GQCP::RSpinOrbitalBasis<double, GQCP::GTOShell> spinor_basis {molecule, "6-31G"};
-    const auto S = spinor_basis.overlap().parameters();
+    const GQCP::USpinOrbitalBasis<double, GQCP::GTOShell> spinor_basis {molecule, "6-31G"};
+    const auto S = spinor_basis.overlap().alpha().parameters();
```
```suggestion
    const auto S = spinor_basis.overlap().alpha().parameters();  // The alpha and beta overlap matrices are equal.
```

lelemmen commented on gqcp/tests/QCMethod/HF/UHF/QCMethod_UHF_test.cpp:
```diff
@@ -56,10 +56,10 @@ BOOST_AUTO_TEST_CASE(h2o_sto3g_plain) {
+    const auto sq_hamiltonian = GQCP::USQHamiltonian<double>::Molecular(spinor_basis, water);  // in an AO basis
 
-    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_alpha, N_beta, sq_hamiltonian, spinor_basis.overlap().parameters());
+    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_alpha, N_beta, sq_hamiltonian, spinor_basis.overlap().alpha().parameters());
```
```suggestion
    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_alpha, N_beta, sq_hamiltonian, spinor_basis.overlap().alpha().parameters());  // The alpha and beta overlap matrices are equal.
```

lelemmen commented on gqcp/tests/QCMethod/HF/UHF/QCMethod_UHF_test.cpp:
```diff
@@ -104,10 +104,10 @@ BOOST_AUTO_TEST_CASE(h2o_sto3g_diis) {
+    const auto sq_hamiltonian = GQCP::USQHamiltonian<double>::Molecular(spinor_basis, water);  // in an AO basis
 
-    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_alpha, N_beta, sq_hamiltonian, spinor_basis.overlap().parameters());
+    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_alpha, N_beta, sq_hamiltonian, spinor_basis.overlap().alpha().parameters());
```
```suggestion
    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_alpha, N_beta, sq_hamiltonian, spinor_basis.overlap().alpha().parameters());  // The alpha and beta overlap matrices are equal.
```

lelemmen commented on gqcp/tests/QCModel/HF/UHF_test.cpp:
```diff
@@ -106,23 +106,22 @@ BOOST_AUTO_TEST_CASE(UHF_DMs) {
     const auto N_a = molecule.numberOfElectronPairs();  // The number of alpha electrons.
     const auto N_b = molecule.numberOfElectronPairs();  // The number of beta electrons.
-    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_a, N_b, r_hamiltonian, spin_orbital_basis.overlap().parameters());
+    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_a, N_b, hamiltonian, u_spin_orbital_basis.overlap().alpha().parameters());
```
```suggestion
    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_a, N_b, hamiltonian, u_spin_orbital_basis.overlap().alpha().parameters());  // The alpha and beta overlap matrices are equal.
```

lelemmen reviewed (approved):
I see now that the UHF module (before this PR) was apparently neglected in the R/U/G refactors #742. I've also opened #845 to address another inconsistency.

Despite the many "suggestions", this PR is great! My eye caught some small documentation that wasn't adjusted for our new style, so I've proposed the changes.

xdvriend commented:
> I see now that the UHF module (before this PR) was apparently neglected in the R/U/G refactors #742. I've also opened #845 to address another inconsistency.

Yes, I saw it. I'll adjust that in this PR as well. Since I'm adding something extra, I'll be sure to re-request a review once it's finished.

> Despite the many "suggestions", this PR is great! My eye caught some small documentation that wasn't adjusted for our new style, so I've proposed the changes.

Ah good catch! I didn't think of fixing the outdated documentation, so thanks for that!

xdvriend commented:
@lelemmen I have re-requested your review on this PR. I changed quite a bit since the last time, so here's a summary:

- I fixed a lot (hopefully all) of the documentation in the HF `QCModel`s and `QCMethod`s.
- I fixed the inconsistency with S, each HF method now uses its respective S-operator.
- I made it so the UHF Environment only contains spin resolved objects.
- I adjusted the python bindings to accommodate these changes.
- I updated the notebooks to use the new python bindings.

Some of the python bindings in the HF `QCMethods` can be seriously revamped in my opinion. I was thinking on tackling that refactor when I implement the RHF and UHF complex bindings.

lelemmen commented on gqcp/include/QCMethod/HF/GHF/GHFDensityMatrixCalculation.hpp:
```diff
@@ -55,11 +55,13 @@ class GHFDensityMatrixCalculation:
      */
     void execute(Environment& environment) override {
+
         const auto& C = environment.coefficient_matrices.back();  // the most recent coefficient matrix
```
```suggestion
        const auto& C = environment.coefficient_matrices.back();  // The most recent coefficient matrix.
```

lelemmen commented on gqcp/include/QCMethod/HF/GHF/GHFFockMatrixCalculation.hpp:
```diff
@@ -56,11 +56,13 @@ class GHFFockMatrixCalculation:
      */
     void execute(Environment& environment) override {
+
         const auto& P = environment.density_matrices.back();  // the most recent density matrix
```
```suggestion
        const auto& P = environment.density_matrices.back();  // The most recent density matrix.
```

lelemmen commented on gqcp/include/QCMethod/HF/UHF/UHFFockMatrixDIIS.hpp:
```diff
@@ -107,13 +106,21 @@ class UHFFockMatrixDIIS:
-        const auto F_alpha_accelerated = this->diis.accelerate(alpha_fock_matrices, error_vectors_alpha);
-        const auto F_beta_accelerated = this->diis.accelerate(beta_fock_matrices, error_vectors_alpha);
+        const auto F_alpha_accelerated = this->diis.accelerate(alpha_fock_matrices, alpha_error_vectors);
+        const auto F_beta_accelerated = this->diis.accelerate(beta_fock_matrices, beta_error_vectors);
```
Yikes! Thanks for fixing this (unnoticed) bug.

lelemmen commented on gqcp/tests/QCMethod/HF/UHF/QCMethod_UHF_stability_test.cpp:
```diff
@@ -39,7 +39,7 @@ BOOST_AUTO_TEST_CASE(H3_stability_test) {
 
     const GQCP::USpinOrbitalBasis<double, GQCP::GTOShell> spinor_basis {molecule, "STO-3G"};
-    const auto S = spinor_basis.overlap().alpha().parameters();
+    const auto S = spinor_basis.overlap();  // The alpha and beta overlap matrices are equal.
```
```suggestion
    const auto S = spinor_basis.overlap();
```

I think I initially suggested this documentation in order to explain why we only took the 'alpha' overlap matrix. Since the API has changed, this isn't necessary anymore!

lelemmen commented on gqcp/tests/QCMethod/HF/UHF/QCMethod_UHF_stability_test.cpp:
```diff
@@ -86,7 +86,7 @@ BOOST_AUTO_TEST_CASE(H4_stability_test) {
 
     const GQCP::USpinOrbitalBasis<double, GQCP::GTOShell> spinor_basis {molecule, "6-31G"};
-    const auto S = spinor_basis.overlap().alpha().parameters();
+    const auto S = spinor_basis.overlap();  // The alpha and beta overlap matrices are equal.
```
```suggestion
    const auto S = spinor_basis.overlap();
```

lelemmen commented on gqcp/tests/QCMethod/HF/UHF/QCMethod_UHF_test.cpp:
```diff
@@ -57,19 +57,19 @@ BOOST_AUTO_TEST_CASE(h2o_sto3g_plain) {
+    const auto sq_hamiltonian = GQCP::USQHamiltonian<double>::Molecular(spinor_basis, water);  // In an AO basis.
 
-    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_alpha, N_beta, sq_hamiltonian, spinor_basis.overlap().alpha().parameters());
+    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_alpha, N_beta, sq_hamiltonian, spinor_basis.overlap());  // The alpha and beta overlap matrices are equal.
```
```suggestion
    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_alpha, N_beta, sq_hamiltonian, spinor_basis.overlap());
```

lelemmen commented on gqcp/tests/QCMethod/HF/UHF/QCMethod_UHF_test.cpp:
```diff
@@ -105,19 +105,19 @@ BOOST_AUTO_TEST_CASE(h2o_sto3g_diis) {
+    const auto sq_hamiltonian = GQCP::USQHamiltonian<double>::Molecular(spinor_basis, water);  // In an AO basis.
 
-    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_alpha, N_beta, sq_hamiltonian, spinor_basis.overlap().alpha().parameters());
+    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_alpha, N_beta, sq_hamiltonian, spinor_basis.overlap());  // The alpha and beta overlap matrices are equal.
```
```suggestion
    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_alpha, N_beta, sq_hamiltonian, spinor_basis.overlap());
```

lelemmen commented on gqcp/tests/QCModel/HF/UHF_test.cpp:
```diff
@@ -109,9 +109,9 @@ BOOST_AUTO_TEST_CASE(UHF_DMs) {
-    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_a, N_b, hamiltonian, u_spin_orbital_basis.overlap().alpha().parameters());
+    const auto N_a = molecule.numberOfElectronPairs();                                                                             // The number of alpha electrons.
+    const auto N_b = molecule.numberOfElectronPairs();                                                                             // The number of beta electrons.
+    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_a, N_b, hamiltonian, u_spin_orbital_basis.overlap());  // The alpha and beta overlap matrices are equal.
```
```suggestion
    auto uhf_environment = GQCP::UHFSCFEnvironment<double>::WithCoreGuess(N_a, N_b, hamiltonian, u_spin_orbital_basis.overlap());
```

lelemmen commented on gqcpy/src/QCMethod/HF/GHF/GHFSCFEnvironment_bindings.cpp:
```diff
@@ -53,9 +53,9 @@ void bindQCModelGHFSCFEnvironmentInterface(Class& py_class) {
             "WithCoreGuess",
-            [](const size_t N, const GSQHamiltonian<Scalar>& hamiltonian, const Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>& S) {
+            [](const size_t N, const GSQHamiltonian<Scalar>& hamiltonian, const ScalarGSQOneElectronOperator<Scalar>& S) {
                 // Use an itermediary Eigen matrix for the Python binding, since Pybind11 doesn't accept our types that are derived from Eigen::Matrix.
```
```suggestion
```

lelemmen commented on gqcpy/src/QCMethod/HF/GHF/GHFSCFEnvironment_bindings.cpp:
```diff
@@ -75,9 +75,9 @@ void bindQCModelGHFSCFEnvironmentInterface(Class& py_class) {
             },
-            [](GHFSCFEnvironment<Scalar>& environment, const Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>& S) {
+            [](GHFSCFEnvironment<Scalar>& environment, ScalarGSQOneElectronOperator<Scalar>& S) {
                 // Use an itermediary Eigen matrix for the Python binding, since Pybind11 doesn't accept our types that are derived from Eigen::Matrix.
```
```suggestion
```

lelemmen commented on gqcpy/src/QCMethod/HF/GHF/GHFSCFEnvironment_bindings.cpp:
```diff
@@ -75,9 +75,9 @@ void bindQCModelGHFSCFEnvironmentInterface(Class& py_class) {
+            [](GHFSCFEnvironment<Scalar>& environment, ScalarGSQOneElectronOperator<Scalar>& S) {
                 // Use an itermediary Eigen matrix for the Python binding, since Pybind11 doesn't accept our types that are derived from Eigen::Matrix.
-                environment.S = SquareMatrix<Scalar>(S);
+                environment.S = ScalarGSQOneElectronOperator<Scalar>(S);
```
```suggestion
                environment.S = S;
```

lelemmen commented on gqcpy/src/QCMethod/HF/RHF/RHFSCFEnvironment_bindings.cpp:
```diff
@@ -55,7 +55,7 @@ void bindRHFSCFEnvironment(py::module& module) {
                 return environment.S;
             },
             [](RHFSCFEnvironment<double>& environment, const Eigen::MatrixXd& S) {
```
```suggestion
            [](RHFSCFEnvironment<double>& environment, const ScalarRSQOneElectronOperator<double>& S) {
```

lelemmen commented on gqcpy/src/QCMethod/HF/RHF/RHFSCFEnvironment_bindings.cpp:
```diff
@@ -55,7 +55,7 @@ void bindRHFSCFEnvironment(py::module& module) {
             },
             [](RHFSCFEnvironment<double>& environment, const Eigen::MatrixXd& S) {
-                environment.S = SquareMatrix<double>(S);
+                environment.S = ScalarRSQOneElectronOperator<double>(S);
```
```suggestion
                environment.S = S;
```

lelemmen commented on gqcp/include/QCMethod/HF/UHF/UHFSCFEnvironment.hpp:
```diff
@@ -78,16 +75,15 @@ class UHFSCFEnvironment {
-    UHFSCFEnvironment(const size_t N_alpha, const size_t N_beta, const USQHamiltonian<Scalar>& sq_hamiltonian, const SquareMatrix<Scalar>& S, const UTransformationComponent<Scalar>& C_alpha_initial, const UTransformationComponent<Scalar>& C_beta_initial) :
-        N_alpha {N_alpha},
-        N_beta {N_beta},
+    UHFSCFEnvironment(const size_t N_alpha, const size_t N_beta, const USQHamiltonian<Scalar>& sq_hamiltonian, const ScalarUSQOneElectronOperator<Scalar>& S, const UTransformationComponent<Scalar>& C_alpha_initial, const UTransformationComponent<Scalar>& C_beta_initial) :
```
Perhaps we could also 
```suggestion
    UHFSCFEnvironment(const size_t N_alpha, const size_t N_beta, const USQHamiltonian<Scalar>& sq_hamiltonian, const ScalarUSQOneElectronOperator<Scalar>& S, const UTransformation<Scalar>& C_initial) :
```

Perhaps we could also change from the 2 arguments to the 1 encapsulating argument? (This also applies to the Python bindings.)

lelemmen commented on gqcpy/src/QCMethod/HF/UHF/UHFSCFEnvironment_bindings.cpp:
```diff
@@ -56,28 +56,26 @@ void bindUHFSCFEnvironment(py::module& module) {
-            [](UHFSCFEnvironment<double>& environment, const Eigen::MatrixXd& S) {
-                environment.S = SquareMatrix<double>(S);
+            [](UHFSCFEnvironment<double>& environment, const ScalarUSQOneElectronOperator<double>& S) {
+                environment.S = ScalarUSQOneElectronOperator<double>(S);
```
```suggestion
                environment.S = S;
```

lelemmen commented on gqcpy/src/QCMethod/HF/UHF/UHFSCFEnvironment_bindings.cpp:
```diff
@@ -56,28 +56,26 @@ void bindUHFSCFEnvironment(py::module& module) {
 
-        .def_readwrite("orbital_energies_alpha", &UHFSCFEnvironment<double>::orbital_energies_alpha)
-        .def_readwrite("orbital_energies_beta", &UHFSCFEnvironment<double>::orbital_energies_beta)
+        .def_readwrite("orbital_energies", &UHFSCFEnvironment<double>::orbital_energies)
```
Does this also require a Python binding for `SpinResolved<...>`?

xdvriend replied:
It does, but I believe `SpinResolved<VectorX<double>>` already has a binding. I could be mistaken and I'll be sure to check before merging this PR!

lelemmen replied:
Ok, I didn't remember.

lelemmen commented on gqcpy/src/QCMethod/HF/UHF/UHFSCFEnvironment_bindings.cpp:
```diff
@@ -204,14 +199,22 @@ void bindUHFSCFEnvironment(py::module& module) {
              [](UHFSCFEnvironment<double>& environment, const Eigen::MatrixXd& new_error_vectors_alpha) {
-                 environment.error_vectors_alpha.pop_back();
-                 environment.error_vectors_alpha.push_back(SquareMatrix<double>(new_error_vectors_alpha));
+                 const auto last_spinResolved_error = environment.error_vectors.back();
+                 const auto last_error_vector_beta = last_spinResolved_error.beta();
+                 SpinResolved<VectorX<double>> new_error {new_error_vectors_alpha, last_error_vector_beta};
+
+                 environment.error_vectors.pop_back();
+                 environment.error_vectors.push_back(new_error);
              })
 
         .def("replace_current_error_vectors_beta",
              [](UHFSCFEnvironment<double>& environment, const Eigen::MatrixXd& new_error_vectors_beta) {
-                 environment.error_vectors_beta.pop_back();
-                 environment.error_vectors_beta.push_back(SquareMatrix<double>(new_error_vectors_beta));
+                 const auto last_spinResolved_error = environment.error_vectors.back();
+                 const auto last_error_vector_alpha = last_spinResolved_error.alpha();
+                 SpinResolved<VectorX<double>> new_error {last_error_vector_alpha, new_error_vectors_beta};
+
+                 environment.error_vectors.pop_back();
+                 environment.error_vectors.push_back(new_error);
              });
```
In line of the spin-resolved types: I would suggest to change this into only providing `.def("replace_current_error_vectors")`. I think it's OK to ask the user to replace either (or both) of the `SpinResolved<...>` components and then push these changes to the environment.

lelemmen reviewed (approved):
Thanks for the additional work that happened in this PR. It's very much appreciated! I'm aware that updating documentation is a chore, and I'm also very happy with the 'unrestricted' spin-resolved changes). The updated files look really fresh!👍🏻 

I've added a couple more suggestions, but I trust that you can resolve them without an additional review. Great job!!!

xdvriend commented:
@lelemmen, I have adjusted the `replace_`API's in the python bindings in order to take two arguments and replace "all" matrices/parameters of that instance. 

Since you mentioned on MS teams that you wanted to review that, I'll leave this PR unmarked until you do. The examples are removed due to a conflict with the develop branch, But I will update the Hartree Fock notebooks with the new bindings in my next PR.

lelemmen commented on gqcp/include/QCMethod/HF/UHF/UHFSCFEnvironment.hpp:
```diff
@@ -99,7 +99,7 @@ class UHFSCFEnvironment {
         UHFSCFEnvironment(rhf_parameters.numberOfElectrons(Spin::alpha), rhf_parameters.numberOfElectrons(Spin::beta),
                           sq_hamiltonian, S,
-                          rhf_parameters.expansion().matrix(), rhf_parameters.expansion().matrix()) {}
+                          UTransformation<Scalar> {rhf_parameters.expansion().matrix(), rhf_parameters.expansion().matrix()}) {}
```
There's a short-hand named constructor for this:
```suggestion
                          UTransformation<Scalar>::FromEqual(rhf_parameters.expansion().matrix())) {}
```

lelemmen commented on gqcpy/src/QCMethod/HF/UHF/UHFSCFEnvironment_bindings.cpp:
```diff
@@ -35,15 +35,14 @@ void bindUHFSCFEnvironment(py::module& module) {
-        .def(py::init([](const size_t N_alpha, const size_t N_beta, const USQHamiltonian<double>& sq_hamiltonian, const ScalarUSQOneElectronOperator<double>& S, const Eigen::MatrixXd& C_alpha_initial, const Eigen::MatrixXd& C_beta_initial) {
-                 return UHFSCFEnvironment<double>(N_alpha, N_beta, sq_hamiltonian, S, UTransformationComponent<double>(C_alpha_initial), UTransformationComponent<double>(C_beta_initial));
+        .def(py::init([](const size_t N_alpha, const size_t N_beta, const USQHamiltonian<double>& sq_hamiltonian, const ScalarUSQOneElectronOperator<double>& S, const UTransformation<double>& C_initial) {
+                 return UHFSCFEnvironment<double>(N_alpha, N_beta, sq_hamiltonian, S, UTransformation<double>(C_initial));
```
```suggestion
                 return UHFSCFEnvironment<double>(N_alpha, N_beta, sq_hamiltonian, S, C_initial);
```

lelemmen reviewed (approved):
I'm thinking of something: didn't you also update some notebooks for updated Python bindings? Did the removal from the `gqcpy/examples/` folder cause any loss of implementations?

Thanks for the additional maintenance work during this PR: the UHF environment's bindings look a lot better in my opinion. Good job!

xdvriend commented:
> I'm thinking of something: didn't you also update some notebooks for updated Python bindings? Did the removal from the `gqcpy/examples/` folder cause any loss of implementations?

Some updates I provided did indeed get removed, but I saved the updated notebooks. So we didn't lose any implementations. It will take me almost no time to fix these notebooks in a subsequent PR. :smile:

> Thanks for the additional maintenance work during this PR: the UHF environment's bindings look a lot better in my opinion. Good job!

Thank you & no problem!
